mmgen/models/diffusions/base_diffusion.py

```diff
@@ -0,0 +1,903 @@
+                [bz, ch, H, W].
+            t (torch.tensor): Target timestep, shape as [bz, ].
+            var (bool, optional): If set as ``True``, would return a dict
+                contains ``var``. Otherwise, only mean would be returned,
+                ``logvar`` would be ignored. Defaults to True.
+            logvar (bool, optional): If set as ``True``, the returned dict
+                would additional contain ``logvar``. This argument would be
+                considered only if ``var == True``. Defaults to False.
+
+        Returns:
+            torch.Tensor | dict: If ``var``, would return a dict contains
+                ``mean`` and ``var``. Otherwise, only mean would be returned.
+                If ``var`` and ``logvar`` set at as True simultaneously, the
+                returned dict would additional contain ``logvar``.
+        """
+
+        tar_shape = x_0.shape
+        tilde_mu_t_coef1 = var_to_tensor(self.tilde_mu_t_coef1, t, tar_shape)
+        tilde_mu_t_coef2 = var_to_tensor(self.tilde_mu_t_coef2, t, tar_shape)
+        posterior_mean = tilde_mu_t_coef1 * x_0 + tilde_mu_t_coef2 * x_t
+        # do not need variance, just return mean
+        if not var:
+            return posterior_mean
+        posterior_var = var_to_tensor(self.tilde_betas_t, t, tar_shape)
+        out_dict = dict(
+            mean_posterior=posterior_mean, var_posterior=posterior_var)
+        if logvar:
+            posterior_logvar = var_to_tensor(self.log_tilde_betas_t_clipped, t,
+                                             tar_shape)
+            out_dict['logvar_posterior'] = posterior_logvar
+        return out_dict
+
+    def p_mean_variance(self,
+                        denoising_output,
+                        x_t,
+                        t,
+                        clip_denoised=True,
+                        denoised_fn=None):
+        r"""Get mean, variance, log variance of denoising process
+        `p(x_{t-1} | x_{t})` and predicted `x_0`.
+
+        Args:
+            denoising_output (dict[torch.Tensor]): The output from denoising
+                model.
+            x_t (torch.Tensor): Diffused image at timestep `t` to denoising.
+            t (torch.Tensor): Current timestep.
+            clip_denoised (bool, optional): Whether cliped sample results into
+                [-1, 1]. Defaults to True.
+            denoised_fn (callable, optional): If not None, a function which
+                applies to the predicted ``x_0`` prediction before it is used
+                to sample. Applies before ``clip_denoised``. Defaults to None.
+
+        Returns:
+            dict: A dict contains ``var_pred``, ``logvar_pred``, ``mean_pred``
+                and ``x_0_pred``.
+        """
+        tar_shape = x_t.shape
+        # prepare for var and logvar
+        if self.denoising_var.upper() == 'LEARNED':
+            # TODO: maybe change to LEARNED_LOG_VAR
```

Comment:
Please finish the TODO items.

Reply:
This `TODO` is a suggestion to change the keyword `LEARNED` to `LEARNED_LOG_VAR` for `self.denoising_var`.
`LEARNED` is used in the official code, but actually, the output is `log_variance`.